vendor/bat-native-ledger/src/bat/ledger/internal/media/reddit.cc

```diff
@@ -0,0 +1,289 @@
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/media/helper.h"
+#include "bat/ledger/internal/media/reddit.h"
+#include "net/http/http_status_code.h"
+#include "url/url_canon.h"
+#include "url/gurl.h"
+
+using std::placeholders::_1;
+using std::placeholders::_2;
+using std::placeholders::_3;
+
+namespace braveledger_media {
+
+MediaReddit::MediaReddit(bat_ledger::LedgerImpl* ledger): ledger_(ledger) {
+}
+
+MediaReddit::~MediaReddit() {
+}
+
+void MediaReddit::ProcessActivityFromUrl(
+    uint64_t window_id,
+    const ledger::VisitData& visit_data) {
+  if (visit_data.path.find("/user/") != std::string::npos) {
+    UserPath(window_id, visit_data);
+    return;
+  }
+  OnMediaActivityError(visit_data, window_id);
+}
+
+void MediaReddit::OnMediaActivityError(
+    const ledger::VisitData& visit_data,
+    uint64_t window_id) {
+  std::string url = REDDIT_TLD;
+  std::string name = REDDIT_MEDIA_TYPE;
+
+  if (!url.empty()) {
+    ledger::VisitData new_visit_data;
+    new_visit_data.domain = url;
+    new_visit_data.url = "https://" + url;
+    new_visit_data.path = "/";
+    new_visit_data.name = name;
+
+    ledger_->GetPublisherActivityFromUrl(
+        window_id, new_visit_data, std::string());
+  }
+}
+
+void MediaReddit::UserPath(
+    uint64_t window_id,
+    const ledger::VisitData& visit_data) {
+  std::string user = GetUserNameFromUrl(visit_data.path);
+
+  if (user.empty()) {
+    OnMediaActivityError(visit_data, window_id);
+    return;
+  }
+
+  std::string media_key = (std::string)REDDIT_MEDIA_TYPE + "_" + user;
+  ledger_->GetMediaPublisherInfo(media_key,
+      std::bind(&MediaReddit::OnUserActivity,
+          this,
+          window_id,
+          visit_data,
+          media_key,
+          _1,
+          _2));
+}
+
+void MediaReddit::OnUserActivity(
+    uint64_t window_id,
+    const ledger::VisitData& visit_data,
+    const std::string& media_key,
+    ledger::Result result,
+    ledger::PublisherInfoPtr publisher_info) {
+  if (!publisher_info || result == ledger::Result::NOT_FOUND) {
+    const std::string user_name = GetUserNameFromUrl(visit_data.path);
+    const std::string url = GetProfileUrl(user_name);
+    FetchDataFromUrl(visit_data.url,
+        std::bind(&MediaReddit::OnUserPage,
+            this,
+            window_id,
+            visit_data,
+            _1,
+            _2,
+            _3));
+  } else {
+    GetPublisherPanelInfo(
+        window_id,
+        visit_data,
+        publisher_info->id);
+  }
+}
+
+void MediaReddit::FetchDataFromUrl(
+    const std::string& url,
+    braveledger_media::FetchDataFromUrlCallback callback) {
+  /* if user is on old reddit, sub the url to get the icon
+     since old reddit didn't have user icons */
+  GURL reddit_url(url);
+  if (reddit_url.DomainIs(OLD_REDDIT_DOMAIN)) {
+    // Canonicalize away 'old.reddit.com' and replace with 'www.reddit.com'.
+    std::string new_host = reddit_url.host().substr(
```

Comment:
Can't you just do a direct assignment here?

`std::string new_host = reddit_url.host();`

Reply:
Yes.

Thanks for the catches.